File: source/frontend/src/lib/dto/response/aggregate/AlternativeAggregateResponseDto.ts
import {ForcedByResponseDto, fromForcedByListResponseDto} from "$lib/dto/response/relationship/ForcedByResponseDto";
import {
    ArchitectureRationaleResponseDto,
    fromArchitectureRationalesResponseDto
} from "$lib/dto/response/entity/ArchitectureRationaleResponseDto";
import {AlternativeResponseDto, fromAlternativeResponseDto} from "$lib/dto/response/entity/AlternativeResponseDto";
import {AlternativeAggregate} from "$lib/domain/aggregate/AlternativeAggregate.svelte";

export class AlternativeAggregateResponseDto {
    alternative : AlternativeResponseDto = AlternativeResponseDto.create();
    forcedBy: ForcedByResponseDto[] = [];
    architectureRationales: ArchitectureRationaleResponseDto[] = [];

    constructor() {}
}

export function fromAlternativeAggregatesResponseDto(alternativeAggregatesResponseDto : AlternativeAggregateResponseDto[]) : AlternativeAggregate[] {
    return alternativeAggregatesResponseDto.map(fromAlternativeAggregateResponseDto);

}

export function fromAlternativeAggregateResponseDto(alternativeAggregateResponseDto : AlternativeAggregateResponseDto) : AlternativeAggregate {
    const alternativeAggregate : AlternativeAggregate = AlternativeAggregate.create();

    alternativeAggregate.alternative = fromAlternativeResponseDto(alternativeAggregateResponseDto.alternative);

    alternativeAggregate.forcedBy = fromForcedByListResponseDto(alternativeAggregateResponseDto.forcedBy);

    alternativeAggregate.architectureRationales = fromArchitectureRationalesResponseDto(alternativeAggregateResponseDto.architectureRationales);

    return alternativeAggregate;
}

File: source/backend/src/main/java/ch/unisg/backend/controller/http/dto/response/ad/AlternativeAggregateResponseDto.java
package ch.unisg.backend.controller.http.dto.response.ad;

import ch.unisg.backend.controller.http.dto.response.ResponseDto;
import ch.unisg.backend.controller.http.dto.response.ar.ArchitecturalRequirementsResponseDto;
import ch.unisg.backend.controller.http.dto.response.ar.ForcedByResponseDto;
import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.UUID;

@Getter @Setter
public class AlternativeAggregateResponseDto extends ResponseDto {

    private UUID id;
    private String title;
    private ForcedByResponseDto[] forcedBy;
    private ArchitectureRationaleResponseDto[] architectureRationales;

    private AlternativeAggregateResponseDto() {

    }

    static public AlternativeAggregateResponseDto create() {
        return new AlternativeAggregateResponseDto();
    }

    public HashMap<String, Object> toJson() {
        HashMap<String, Object> alternativeResponseDto = new HashMap<>();

        alternativeResponseDto.put("id", this.id);
        alternativeResponseDto.put("title", this.title);
        alternativeResponseDto.put("forcedBy", this.forcedBy);
        alternativeResponseDto.put("architectureRationale", this.architectureRationales);

        return alternativeResponseDto;

    }

    public static HashMap<String, Object> toJsonVerbose(AlternativeAggregate alternativeAggregate) {

        HashMap<String, Object> alternative = new HashMap<>();

        alternative.put("alternative", AlternativeResponseDto.toJson(alternativeAggregate.getAlternative()));
        alternative.put("forcedBy", ArchitecturalRequirementsResponseDto.toJson(alternativeAggregate.getForcedBy()));
        alternative.put("architectureRationales", ArchitectureRationaleResponseDto.toJson(alternativeAggregate.getArchitectureRationales()));

        return alternative;

    }

    public static List<HashMap<String, Object>> toJsonVerbose(List<AlternativeAggregate> alternativeAggregates) {

        List<HashMap<String, Object>> alternatives = new ArrayList<>();

        alternativeAggregates.forEach(alternativeAggregate -> {
            alternatives.add(toJsonVerbose(alternativeAggregate));
        });

        return alternatives;

    }

}
